File: manhunt-base/src/main/java/io/github/ytg1234/manhunt/base/init/ManhuntTicks.java
package io.github.ytg1234.manhunt.base.init;

import io.github.ytg1234.manhunt.api.event.callback.SpeedrunnerGlowCallback;
import io.github.ytg1234.manhunt.base.ManhuntUtils;
import io.github.ytg1234.manhunt.base.config.ManhuntConfigEnums;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.server.MinecraftServer;

import static io.github.ytg1234.manhunt.base.ManhuntUtils.fromServer;

public final class ManhuntTicks {
    public static void centralTick(MinecraftServer server) {
        updateCompasses(server);
        highlightSpeedrunner(server);
    }

    private static void updateCompasses(MinecraftServer server) {
        ManhuntUtils.hunters.forEach(hunterUuid -> {
            // Check if player is null
            if (fromServer(server, hunterUuid) == null) return;

            // Continue with logic
            ItemStack stack = fromServer(server, hunterUuid).inventory.getStack(8);

            // If the stack is empty, null or not a compass put a compass there
            if (stack == null || stack.isEmpty() || !stack.getItem().equals(Items.COMPASS)) {
                if (ManhuntUtils.CONFIG.giveCompassWhenSettingHunters) {
                    fromServer(server, hunterUuid).equip(8, new ItemStack(Items.COMPASS));
                    stack = fromServer(server, hunterUuid).inventory.getStack(8);
                } else if (stack == null) return;
            }
            if (ManhuntUtils.CONFIG.compassBehaviour.equals(ManhuntConfigEnums.Compass.UPDATE)) {
                // Set compass NBT
                if (stack.getItem().equals(Items.COMPASS)) {
                    fromServer(server, hunterUuid).equip(8, ManhuntUtils.updateCompass(stack, fromServer(server, ManhuntUtils.speedrunner)));
                }
            }
        });
    }

    private static void highlightSpeedrunner(MinecraftServer server) {
        // If speedrunner is null, bad.
        if (fromServer(server, ManhuntUtils.speedrunner) == null) return;
        if (ManhuntUtils.CONFIG.highlightSpeedrunner) {
            boolean toCancel = SpeedrunnerGlowCallback.EVENT.invoker().onSpeedrunnerGlow(fromServer(server, ManhuntUtils.speedrunner));
            if (!toCancel) ManhuntUtils.applyStatusEffectToPlayer(fromServer(server, ManhuntUtils.speedrunner), StatusEffects.GLOWING);
        }
    }
}

File: manhunt-base/src/main/kotlin/io/github/ytg1234/manhunt/base/init/ManhuntTicks.kt
package io.github.ytg1234.manhunt.base.init

import io.github.ytg1234.manhunt.api.event.callback.SpeedrunnerGlowCallback
import io.github.ytg1234.manhunt.base.*
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.server.MinecraftServer

/**
 * Tick tock.
 *
 * @author YTG1234
 */
object ManhuntTicks {
    /**
     * Activates all tick events.
     *
     * @see net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents.EndTick.onEndTick
     */
    fun centralTick(server: MinecraftServer) {
        updateCompasses(server)
        highlightSpeedrunner(server)
    }

    /**
     * Updates the compasses of each hunter.
     *
     * @see net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents.EndTick.onEndTick
     */
    private fun updateCompasses(server: MinecraftServer) {
        hunters.forEach { hunterUuid ->
            // Continue with logic
            if (fromServer(server, hunterUuid) == null) return@forEach
            var stack: ItemStack = fromServer(server, hunterUuid)!!.inventory.getStack(8)

            // If the stack is empty, null or not a compass put a compass there
            if (stack.isEmpty || stack.item != Items.COMPASS) {
//                if (CONFIG!!.giveCompassWhenSettingHunters) {
                    fromServer(server, hunterUuid)!!.inventory.setStack(8, ItemStack(Items.COMPASS))
                    stack = fromServer(server, hunterUuid)!!.inventory.getStack(8)
//                }
            }
//            if (CONFIG!!.compassBehaviour == Compass.UPDATE) {
//                // Set compass NBT
//                if (stack.item == Items.COMPASS) {
//                    fromServer(server, hunterUuid)!!.inventory.setStack(
//                        8,
//                        updateCompass(stack, fromServer(server, speedrunner))
//                    )
//                }
//            }
        }
    }

    /**
     * Applies the glowing effect to the speedrunner
     * if the configuration allows that.
     *
     * @see net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents.EndTick.onEndTick
     */
    private fun highlightSpeedrunner(server: MinecraftServer) {
        // If speedrunner is null, bad.
//        if (CONFIG!!.highlightSpeedrunner) {
//            val toCancel =
//                SpeedrunnerGlowCallback.EVENT.invoker().onSpeedrunnerGlow(fromServer(server, speedrunner))
//            if (!toCancel && speedrunner != null) applyStatusEffectToPlayer(
//                fromServer(server, speedrunner)!!,
//                StatusEffects.GLOWING
//            )
//        }
    }
}
